botocore/handlers.py

```diff
@@ -450,6 +450,38 @@ def document_base64_encoding(param):
     return append.append_documentation
 
 
+def validate_ascii_metadata(params, **kwargs):
+    """Verify S3 Metadata only contains ascii characters.
+
+    From: http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html
+
+    "Amazon S3 stores user-defined metadata in lowercase. Each name, value pair
+    must conform to US-ASCII when using REST and UTF-8 when using SOAP or
+    browser-based uploads via POST."
+
+    """
+    metadata = params.get('Metadata')
+    if not metadata or not isinstance(metadata, dict):
```

Comment:
Yeah that is an interesting problem in that we cannot validate with the validator that the parameter passed by the user is correct prior to the handler. It would be interesting to expose a new event down the line passed the validation or a helper method to run a quick validation in the future if we run into more of these scenarios.